lib/datadog/core/telemetry/event.rb

```diff
@@ -1,81 +1,277 @@
 # frozen_string_literal: true
 
-require_relative 'collector'
-require_relative 'v1/app_event'
-require_relative 'v1/telemetry_request'
-require_relative 'v2/app_client_configuration_change'
-
 module Datadog
   module Core
     module Telemetry
-      # Class defining methods to construct a Telemetry event
       class Event
-        include Telemetry::Collector
+        # Base class for all Telemetry V2 events.
+        class Base
```

Comment:
Should this be a class? Or could it be a module? (Given it doesn't introduce state.)

Reply:
I think the hierarchy here is that events are true children of a base class: an event is a subtype of a `Base` event.

It can technically work as a module, but I think that it would create the misleading impression that events just so happen to respect the `Base` interface, when in fact all events are a subtype of a `Base` event.

Reply:
I see, this got me thinking about Sandi Metz's take on inheritance/composition. Some notes surrounding her discourse on the subject that I could find:

> *Deciding Between Inheritance and Composition*

> Inheritance gives you message delegation for free at the cost of maintaining a class hierarchy. Composition allows objects to have structural independence at the cost of explicit message delegation.

> > Composition contains far few built-in dependencies than inheritance; it is very often the best choice.

> With inheritance, a correctly modeled hierarchy will give you the benefit of propogating changes in the base class to all subclasses. This can also be a disadvantage when the hierarchy is modeled incorrectly, as a dramatic change to the base class due to a change in requirements will break sub-classes. Inheritance = built-in dependencies.

> > Enormous, broad-reaching changes of behavior can be achieved with very small changes in code. This is true, for better or for worse, whether you come to regret it or not.

> > Avoid writing frameworks that require users of your code to subclass your objects in order to gain your behavior. Their application's objects may already be arranged in a hierarchy; inheriting from your framework may not be possible.

On that very last point: is there a case where we might want to have independent trees of event behavior? I don't think immediately, no, but that would be a strong case for composition over inheritance. I suppose to me, it would be more a factor of keeping our options open in accordance to the principle of least harm: that inheritance would be less restraining.

Another way I think about the reason to do one over the other is state (e.g. instance variables).

 - Does the base class define or maintain state? (Internal or otherwise?)
 - Does it need some initialization (constructor or otherwise)?

If yes to either, then probably should be inheritance, because modules are not great at maintaining state. Otherwise consider using a module.

Some food for thought. Ultimately, we can always refactor and this doesn't constitute a breaking change, so no blocker on my side if we can always change later without much pain.

Reply:
> Deciding Between Inheritance and Composition

It seems like the composition that she is discussing is about having a delegator object. And I'm not sure that that applies here because the need is for a consistent interface. There's almost no behaviour being delegated to the base class, is for subclasses to respect interface.

Both classes and modules would fall under inheritance, based on her definition above I believe.

Reply:
Also, this whole thing is internal and constraint inside a single file, so factoring is trivial, and will likely not even affect unit tests.